mlir/lib/Dialect/Rock/Transforms/AccelEmitter.h

```diff
@@ -0,0 +1,197 @@
+
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/Dialect/Rock/IR/MfmaInsnGroup.h"
+#include "mlir/Dialect/Rock/IR/TransformMapBuilder.h"
+#include "mlir/Dialect/Rock/IR/WmmaInsnGroup.h"
+#include "mlir/Dialect/Rock/utility/builderUtils.h"
+#include "mlir/Dialect/Vector/IR/VectorOps.h"
+
+#include <memory>
+
+namespace mlir {
+namespace rock {
+namespace accel {
+struct AccelEmitter {
+
+  // Select the right accelerator based on the set of features and architecture
+  static std::unique_ptr<AccelEmitter>
+  select(GemmFeatures features, Type dataTypeA, Type dataTypeB, StringRef arch,
+         XdlopsGemmParamsAttr tuningParams);
+
+  AccelEmitter(StringRef arch, XdlopsGemmParamsAttr tuningParams);
+  /// Emit the actual intrinsic in the threadwise operation
+  virtual void emitThreadwiseLoop(OpBuilder &b, Location loc, Value argA,
+                                  Value argB, Value bufferC,
+                                  Value regCOffset) = 0;
+
+  // Compute the correct lds source offset when loading data from shared memory
+  // into registers. The pseudo-code of the lds-to-register loops is as follows
+  // for(index_t m_i = 0; m_i < mRepeats; ++m_i)
+  //   for(index_t k_i = 0; k_i < KPerThread; ++k_i)
+  //       sourceOffset = computeLdsSourceOffset(d_i, k_i, dPerBlock,
+  //       baseOffset)
+  //       ...
+  // In the above loop `d` can be either `m` or `n`.
+  virtual Value computeLdsSourceOffset(OpBuilder &kBuilder, Value k_i,
+                                       OpBuilder &dBuilder, Value d_i,
+                                       OpBuilder &builder, Value dPerBlock,
+                                       Location loc, Value baseOffset,
+                                       Value laneId) = 0;
+
+  // Compute the output transform map to be used to store the result of the
+  // matrix multiplication tile
+  virtual ArrayAttr computeOutputTransforms(PatternRewriter &b, Location loc,
+                                            int64_t matrixM, int64_t matrixN,
+                                            int64_t blockSize, int64_t gridSize,
+                                            Value regC) = 0;
+
+  // Convert from memref<?xvector<?xT>> to memref<?xD> where the source T
+  // is the accumulator type and D is the destination type
+  Value computeOutputConversion(PatternRewriter &b, Location loc,
+                                int64_t matrixM, int64_t matrixN,
+                                int64_t blockSize, int64_t gridSize,
+                                Value regDest, Value convertedC,
+                                bool forceUnroll);
+
+  // Validate the accelerator structure
+  void validateAcceleratorProperties();
+
+  virtual ~AccelEmitter() {}
+
+public:
+  //
+  // Tuning parameters
+  //
+  int64_t mPerBlock;
+  int64_t nPerBlock;
+  int64_t kPerBlock;
+  int64_t mPerWave;
+  int64_t nPerWave;
+  int64_t kPack;
+  int64_t waveSize;
```

Comment:
Can we reuse the tuning parameters structure instead of making a copy of every parameter here.

Reply:
Done. I guess it's a matter of preference, because now I have to extract those parameters locally in different functions (it's no big deal, just style)